azkaban-common/src/main/java/azkaban/trigger/Trigger.java

```diff
@@ -120,120 +121,101 @@ public void setContext(Map<String, Object> context) {
+    private Map<String, Object> context = new HashMap<String, Object>();
+
+    public TriggerBuilder(String submitUser,
+                          String source,
+                          Condition triggerCondition,
+                          Condition expireCondition,
+                          List<TriggerAction> actions) {
+      this.submitUser = submitUser;
+      this.source = source;
+      this.triggerCondition = triggerCondition;
+      this.actions = actions;
+      this.expireCondition = expireCondition;
+      long now = DateTime.now().getMillis();
+      this.submitTime = now;
+      this.lastModifyTime = now;
+    }
 
-  public Trigger(String submitUser, String source, Condition triggerCondition,
-      Condition expireCondition, List<TriggerAction> actions,
-      List<TriggerAction> expireActions) {
-    this.lastModifyTime = DateTime.now().getMillis();
-    this.submitTime = DateTime.now().getMillis();
-    this.submitUser = submitUser;
-    this.source = source;
-    this.triggerCondition = triggerCondition;
-    this.expireCondition = expireCondition;
-    this.actions = actions;
-    this.expireActions = expireActions;
-  }
+    public TriggerBuilder setId(int id) {
+      this.triggerId = id;
+      return this;
+    }
 
-  public Trigger(String submitUser, String source, Condition triggerCondition,
-      Condition expireCondition, List<TriggerAction> actions) {
-    this.lastModifyTime = DateTime.now().getMillis();
-    this.submitTime = DateTime.now().getMillis();
-    this.submitUser = submitUser;
-    this.source = source;
-    this.triggerCondition = triggerCondition;
-    this.expireCondition = expireCondition;
-    this.actions = actions;
-    this.expireActions = new ArrayList<TriggerAction>();
-  }
+    public TriggerBuilder setSubmitTime(long time) {
+      this.submitTime = time;
+      return this;
+    }
 
-  public Trigger(long lastModifyTime, long submitTime, String submitUser,
-      String source, Condition triggerCondition, Condition expireCondition,
-      List<TriggerAction> actions) {
-    this.lastModifyTime = lastModifyTime;
-    this.submitTime = submitTime;
-    this.submitUser = submitUser;
-    this.source = source;
-    this.triggerCondition = triggerCondition;
-    this.expireCondition = expireCondition;
-    this.actions = actions;
-    this.expireActions = new ArrayList<TriggerAction>();
+    public TriggerBuilder setLastModifyTime(long time) {
+      this.lastModifyTime = time;
+      return this;
+    }
+
+    public TriggerBuilder setExpireActions(List<TriggerAction> actions) {
+      this.expireActions = actions;
+      return this;
+    }
+
+    public TriggerBuilder setInfo(Map<String, Object> info) {
+      this.info = info;
+      return this;
+    }
+
+    public TriggerBuilder setContext(Map<String, Object> context) {
+      this.context = context;
+      return this;
+    }
+
+    public Trigger build() {
+      return new Trigger(triggerId,
```

Comment:
discussed offline, i removed that private constructor and made fields without associated setters final.